File: app/src/main/java/com/upence/data/SMSParsingPattern.kt
package com.upence.data

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "sms_parsing_patterns")
data class SMSParsingPattern(
    @PrimaryKey(autoGenerate = true)
    val id: Int = 0,
    val senderIdentifier: String, // Phone number or sender ID pattern
    val patternName: String, // User-friendly name for the pattern
    val messageStructure: String, // Simplified structure for matching
    val amountPattern: String, // Regex or position for amount extraction
    val counterpartyPattern: String, // Regex or position for counterparty extraction
    val datePattern: String? = null, // Optional date extraction
    val referencePattern: String? = null, // Optional reference number extraction
    val transactionType: TransactionType, // CREDIT or DEBIT
    val isActive: Boolean = true, // Whether this pattern should be used
    val defaultCategoryID: String = "",
    val defaultAccountID: String = "",
    val autoSelectAccount: Boolean = false,
    val senderName: String = "",
    val createdTimestamp: Long = System.currentTimeMillis(),
    val lastUsedTimestamp: Long? = null
)

File: lib/data/models/sms_parsing_pattern.dart
import 'transaction.dart';

class SMSParsingPattern {
  final int id;
  final String senderIdentifier;
  final String patternName;
  final String messageStructure;
  final String amountPattern;
  final String counterpartyPattern;
  final String? datePattern;
  final String? referencePattern;
  final TransactionType transactionType;
  final bool isActive;
  final String defaultCategoryId;
  final String defaultAccountId;
  final bool autoSelectAccount;
  final String senderName;
  final String sampleSms;
  final int createdTimestamp;
  final int? lastUsedTimestamp;

  SMSParsingPattern({
    this.id = 0,
    required this.senderIdentifier,
    required this.patternName,
    required this.messageStructure,
    required this.amountPattern,
    required this.counterpartyPattern,
    this.datePattern,
    this.referencePattern,
    required this.transactionType,
    this.isActive = true,
    this.defaultCategoryId = '',
    this.defaultAccountId = '',
    this.autoSelectAccount = false,
    this.senderName = '',
    this.sampleSms = '',
    required this.createdTimestamp,
    this.lastUsedTimestamp,
  });

  SMSParsingPattern copyWith({
    int? id,
    String? senderIdentifier,
    String? patternName,
    String? messageStructure,
    String? amountPattern,
    String? counterpartyPattern,
    String? datePattern,
    String? referencePattern,
    TransactionType? transactionType,
    bool? isActive,
    String? defaultCategoryId,
    String? defaultAccountId,
    bool? autoSelectAccount,
    String? senderName,
    String? sampleSms,
    int? createdTimestamp,
    int? lastUsedTimestamp,
  }) {
    return SMSParsingPattern(
      id: id ?? this.id,
      senderIdentifier: senderIdentifier ?? this.senderIdentifier,
      patternName: patternName ?? this.patternName,
      messageStructure: messageStructure ?? this.messageStructure,
      amountPattern: amountPattern ?? this.amountPattern,
      counterpartyPattern: counterpartyPattern ?? this.counterpartyPattern,
      datePattern: datePattern ?? this.datePattern,
      referencePattern: referencePattern ?? this.referencePattern,
      transactionType: transactionType ?? this.transactionType,
      isActive: isActive ?? this.isActive,
      defaultCategoryId: defaultCategoryId ?? this.defaultCategoryId,
      defaultAccountId: defaultAccountId ?? this.defaultAccountId,
      autoSelectAccount: autoSelectAccount ?? this.autoSelectAccount,
      senderName: senderName ?? this.senderName,
      sampleSms: sampleSms ?? this.sampleSms,
      createdTimestamp: createdTimestamp ?? this.createdTimestamp,
      lastUsedTimestamp: lastUsedTimestamp ?? this.lastUsedTimestamp,
    );
  }
}
